openhands/llm/retry_mixin.py

```diff
@@ -0,0 +1,85 @@
+from tenacity import (
+    retry,
+    retry_if_exception_type,
+    stop_after_attempt,
+    wait_exponential,
+)
+
+from openhands.core.exceptions import OperationCancelled
+from openhands.core.logger import openhands_logger as logger
+from openhands.runtime.utils.shutdown_listener import should_exit
+
+
+class RetryMixin:
+    """Mixin class for retry logic."""
+
+    def retry_decorator(self, **kwargs):
+        """
+        Create a retry decorator with customizable parameters.
+
+        Args:
+            **kwargs: Keyword arguments to override default retry behavior.
+                      Keys: num_retries, retry_exceptions, retry_min_wait, retry_max_wait, retry_multiplier
+
+        Returns:
+            A retry decorator with the specified or default parameters.
+        """
+        num_retries = kwargs.get('num_retries')
+        retry_exceptions = kwargs.get('retry_exceptions')
+        retry_min_wait = kwargs.get('retry_min_wait')
+        retry_max_wait = kwargs.get('retry_max_wait')
+        retry_multiplier = kwargs.get('retry_multiplier')
+
+        def custom_completion_wait(self, retry_state):
+            """Custom wait for completion."""
+            exception = retry_state.outcome.exception() if retry_state.outcome else None
+            if exception is None:
+                return 0
```

Comment:
Definitely minor, but it seems strange, why would it retry if it doesn't have an exception?

Reply:
That originated from when it was used with "after", which had this called for both fails and successes.
IF we stay with the before_sleep option, this check could be removed, I think.

Reply:
Definitely, `before_sleep` makes perfect sense.